streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -500,17 +514,43 @@ int maybeCommitActiveTasksPerUserRequested() {
         if (rebalanceInProgress) {
             return -1;
         } else {
-            int commits = 0;
+            final Map<TaskId, Map<TopicPartition, OffsetAndMetadata>> consumedOffsetsAndMetadataPerTask = new HashMap<>();
             for (final Task task : activeTaskIterable()) {
                 if (task.commitRequested() && task.commitNeeded()) {
-                    task.commit();
-                    commits++;
+                    consumedOffsetsAndMetadataPerTask.put(task.id(), task.prepareCommit());
+                    task.markCommitted();
                 }
             }
-            return commits;
+            commit(consumedOffsetsAndMetadataPerTask);
+
+            return consumedOffsetsAndMetadataPerTask.size();
         }
     }
 
+    private void commit(final Map<TaskId, Map<TopicPartition, OffsetAndMetadata>> offsetsPerTask) {
+        if (eosEnabled) {
+            for (final Map.Entry<TaskId, Map<TopicPartition, OffsetAndMetadata>> taskToCommit : offsetsPerTask.entrySet()) {
+                taskProducers.get(taskToCommit.getKey()).commitTransaction(taskToCommit.getValue());
+            }
+        } else {
```

Comment:
What's the reasoning her for only wrapping the consumer offset commit case here, not for EOS case?

Reply:
You mean exception handling? For the producer all exception handling is done within `StreamsProducer` (note that `threadProducer` above is a `StreamsProducer`, not a `KafkaProducer`)